lib/vrl/compiler/src/expression/function_closure.rs

```diff
@@ -0,0 +1,194 @@
+                    .into_inner();
+
+                for (index, value) in ArrayRecursive::create_iter(array, recursive) {
+                    let state = ctx.state_mut();
+                    state.insert_variable(index_ident.clone(), index.into());
+                    state.insert_variable(value_ident.clone(), value);
+
+                    let output = self.block.resolve(ctx)?;
+
+                    result = func(ctx, output, result)?;
+
+                    let state = ctx.state_mut();
+                    state.remove_variable(&index_ident);
+                    state.remove_variable(&value_ident);
+                }
+                Ok(result)
+            }
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl fmt::Display for FunctionClosure {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // TODO
+        f.write_str("{ |")?;
+
+        let mut iter = self.variables.iter().peekable();
+        while let Some(var) = iter.next() {
+            var.fmt(f)?;
+
+            if iter.peek().is_some() {
+                f.write_str(", ")?;
+            }
+        }
+
+        f.write_str("|\n")?;
+        self.block.fmt(f)?;
+
+        f.write_str("\n}")
+    }
+}
+
+struct ObjectRecursive {
+    queue: VecDeque<(String, Value)>,
+}
+
+impl ObjectRecursive {
+    fn new(map: BTreeMap<String, Value>) -> Self {
+        let queue = map.into_iter().collect();
+        Self { queue }
+    }
+
+    fn create_iter(
+        map: BTreeMap<String, Value>,
+        recursive: bool,
+    ) -> Box<dyn Iterator<Item = (String, Value)>> {
+        if recursive {
+            Box::new(ObjectRecursive::new(map).into_iter())
+        } else {
+            Box::new(map.into_iter())
+        }
+    }
+}
+
+impl Iterator for ObjectRecursive {
+    type Item = (String, Value);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (key, value) = self.queue.pop_front()?;
+        match value {
+            Value::Object(ref map) => {
+                self.queue.extend(map.clone().into_iter());
+                Some((key, value))
+            }
+            _ => Some((key, value)),
+        }
+    }
+}
+
+struct ArrayRecursive {
+    queue: VecDeque<(usize, Value)>,
+}
+
+impl ArrayRecursive {
+    fn new(array: Vec<Value>) -> Self {
+        let queue = array.into_iter().enumerate().collect();
+        Self { queue }
+    }
+
+    fn create_iter(array: Vec<Value>, recursive: bool) -> Box<dyn Iterator<Item = (usize, Value)>> {
+        if recursive {
+            Box::new(ArrayRecursive::new(array).into_iter())
+        } else {
+            Box::new(array.into_iter().enumerate())
+        }
+    }
+}
+
+impl Iterator for ArrayRecursive {
+    type Item = (usize, Value);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (index, value) = self.queue.pop_front()?;
+        match value {
+            Value::Array(ref array) => {
+                self.queue.extend(array.clone().into_iter().enumerate());
```

Comment:
Is it reasonable to have the index value for a recursively visited array be based on the given array's indexes?